crates/tests/src/integration/masp.rs

```diff
@@ -7830,3 +8138,521 @@ fn speculative_context() -> Result<()> {
+                validator_one_rpc,
+            ]),
+        )
+    });
+    assert!(captured.result.is_ok());
+
+    let file_path = tempdir
+        .path()
+        .read_dir()
+        .unwrap()
+        .next()
+        .unwrap()
+        .unwrap()
+        .path();
+    txs_bytes.push(std::fs::read(&file_path).unwrap());
+    std::fs::remove_file(&file_path).unwrap();
+
+    // Construct the batch
+    let captured = CapturedOutput::of(|| {
+        run(
+            &node,
+            Bin::Client,
+            apply_use_device(vec![
+                "unshield",
+                "--source",
+                C_SPENDING_KEY,
+                "--target",
+                cooper_alias.as_ref(),
+                "--token",
+                NAM,
+                "--amount",
+                "1",
+                "--gas-limit",
+                "100000",
+                "--gas-price",
+                "0.00001",
+                "--gas-payer",
+                cooper_alias.as_ref(),
+                "--gas-spending-key",
+                C_SPENDING_KEY,
+                "--output-folder-path",
+                tempdir.path().to_str().unwrap(),
+                "--dump-tx",
+                "--ledger-address",
+                validator_one_rpc,
+            ]),
+        )
+    });
+    assert!(captured.result.is_ok());
+
+    let file_path = tempdir
+        .path()
+        .read_dir()
+        .unwrap()
+        .next()
+        .unwrap()
+        .unwrap()
+        .path();
+    let batch_tx0_bytes = std::fs::read(&file_path).unwrap();
+    std::fs::remove_file(&file_path).unwrap();
+
+    let captured = CapturedOutput::of(|| {
+        run(
+            &node,
+            Bin::Client,
+            apply_use_device(vec![
+                "transfer",
+                "--source",
+                C_SPENDING_KEY,
+                "--target",
+                AB_PAYMENT_ADDRESS,
+                "--token",
+                NAM,
+                "--amount",
+                "1",
+                // Fake a transparent gas payer, fees will actually be paid by
+                // the first tx of this batch
+                "--gas-payer",
+                CHRISTEL_KEY,
+                "--output-folder-path",
+                tempdir.path().to_str().unwrap(),
+                "--dump-tx",
+                "--ledger-address",
+                validator_one_rpc,
+            ]),
+        )
+    });
+    assert!(captured.result.is_ok());
+
+    let file_path = tempdir
+        .path()
+        .read_dir()
+        .unwrap()
+        .next()
+        .unwrap()
+        .unwrap()
+        .path();
+    let batch_tx1_bytes = std::fs::read(&file_path).unwrap();
+    std::fs::remove_file(&file_path).unwrap();
+
+    // Create the batch
+    let tx0: namada_sdk::tx::Tx =
+        serde_json::from_slice(&batch_tx0_bytes).unwrap();
+    let tx1: namada_sdk::tx::Tx =
+        serde_json::from_slice(&batch_tx1_bytes).unwrap();
```

Comment:
you can reconstruct the expected commitment tree from these txs, by getting the `MaspTxId` within. from the rebuilt commitment tree, after execution, you can check that the tree root is in namada's storage

Reply:
Good one, I've done it in 0444309cac